### PR TITLE
Functions

### PR DESCRIPTION
**Pull request by Pavel-Durov**
functions support:
- lexing
- parsing
- basic evaluation
- tests

#### Function Declaration

```
fun <name> (<params>) { <statements> }
```

`<fun>` - keyword indicating the beginning of a function declaration

`<name>` - Same as variable names

`<params>` - (optional) single or list of parameters passed to the function

`<statements>` - statements that comprise the body of the function

Example:
```
"fun _add (_arg1, _arg2){ return _arg1 + _arg2; }
```
#### Function calls

`<name>` (`<arguments>`)

`<arguments>` - (optional) single or list of parameters passed to the function

Example:
```
_add(1,2)
```

### CODE REVIEW

**Comment by vext01** [comment]
I'll review this once the other PR is merged and this is rebased.

**Comment by Pavel-Durov** [comment]
@vext01 rebased and ready for review

**Comment by Pavel-Durov** [comment]
bors try

**Comment by Pavel-Durov** [comment]
bors try

**Comment by Pavel-Durov** [comment]
I think I screwed up something in my rebase... Not ready for review yet

**Comment by Pavel-Durov** [comment]
bors try

File: src/err.rs [hunk 1]
```diff
@@ -8,12 +8,20 @@ pub enum InterpError {
...
 
 impl Display for InterpError {
     fn fmt(&self, f: &mut Formatter) -> Result {
         match self {
+            InterpError::EvalError(msg) => {
+                f.write_str(format!("Evaluation error: {}!", msg).as_str())
+            }
             InterpError::EmptyStack => f.write_str("Cannot pop from empty stack!"),
+            InterpError::UndefinedFunction(id) => {
+                f.write_str(format!("Cannot find function wiht id '{}'!", id).as_str())
```
Comment (vext01) [inline src/err.rs:23]: typo here

**Comment by Pavel-Durov** [comment]
Ah, I think you got me there :no_mouth: 

1. There's no support for function block with multiple statements, so:
```
fun _f1(_a) {
    _f2();
    println(_a);
}
```
will fail to parse

2. There's also no support for new lines in blocks

That was intentional to keep the PRs small :)

But If I'm not mistaken, I think your question is regarding function scope. 

If I adapt your example to something that YAIWR can currently execute:
```
fun _f1(_a) {_f2()}
fun _f2() { let _a = 3;}
fun _main() {_f1(1)}
_main()
println(_a);
```
It will print "3" since I have only one GLOBAL scope for all functions.
I started addressing the function scope in my next PR (draft) :point_right: https://github.com/softdevteam/pavel.yaiwr/pull/19

**Comment by vext01** [comment]
>  I have only one GLOBAL scope for all functions.

That's what I was wondering. I agree that you probably do want some notion of local scope. Without it recursion would be hard.

**Comment by vext01** [comment]
Please squash.

**Pull request merged**